js/serializers/scormSuspendDataSerializer.js

```diff
@@ -1,691 +1,1129 @@
-
-          var block = bin.substr(b*6,6);
-          base64 += Base64(parseInt(block, 2));
+    /**
+     * Convert an array, boolean or number into a string binary representation
+     * @param {number|boolean|Array} value
+     * @param {string} [typeName] To force an internal data type for testing
+     * @param {boolean} [logStats] Console logs statistics on type use
+     * @returns {string}
+     */
+    valueToBin(value, typeName = null, logStats = null) {
+      this.clearLog();
+      const hasInitialType = (typeof typeName === 'string');
+      value = convertValuesToNumbers(value);
+      let valueType;
+      if (hasInitialType) {
+        valueType = findValueTypeFromName(typeName);
+        if (!valueType) {
+          throw new Error(`Could not find value type from name "${typeName}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
+        }
+      } else {
+        valueType = findValueTypeFromValue(value);
+        if (!valueType) {
+          throw new Error(`Could not find value type from value "${value}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
         }
+      }
+      const bin = valueType.valueToBin(value, logStats);
+      if (!hasInitialType) {
+        bin.unshift(valueType.binType);
+      }
+      logStats && this.printLog();
+      return bin;
+    }
 
-        return base64;
-      },
-      "base16": function ConverterBinaryToBase16(bin) {
-        var paddingLength = 0;
-        if (bin.length % 4) paddingLength = 4 - (bin.length % 4);
-        binPaddingLen = NumberToBinary(paddingLength, 4);
-        binPadding = Padding.addLeft("", paddingLength);
-        bin = binPaddingLen + binPadding + bin;
-
-        var binLength = bin.length;
-        var hex = "";
-        for (var b = 0; b < 10000; b++) {
-          if (b*4 >= binLength) break;
-
-          var block = bin.substr(b*4,4);
-          hex += parseInt(block, 2).toString(16);
+    /**
+     * Convert the string binary representation back into an array, boolean or number.
+     * @param {string} bin A string binary representation of data from valueToBin
+     * @param {string} [typeName] To force an internal data type for testing
+     * @returns {number|boolean|Array}
+     */
+    valueFromBin(bin, typeName = null) {
+      bin = _.flatten(bin).join('');
+      let valueType;
+      if (typeof typeName === 'string') {
+        valueType = findValueTypeFromName(typeName);
+        if (!valueType) {
+          throw new Error(`Could not find value type from name "${typeName}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
+        }
+      } else {
+        [valueType, bin]= shiftValueTypeFromBin(bin);
+        if (!valueType || !(valueType instanceof AbstractValueType)) {
+          throw new Error(`Could not find value type from value "${value}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
```

Comment:
As per previous comment, could refactor to be DRYer.

There's something dodgy in the existing code anyhow – is `value` meant to be `bin` in the second error?
```suggestion
      const hasInitialType = typeof typeName === 'string';
      const valueType = hasInitialType ?
        findValueTypeFromName(typeName) :
        shiftValueTypeFromBin(bin)[0];
      if (!valueType || !(valueType instanceof AbstractValueType)) {
        throw new Error(`Could not find value type from name "${hasInitialType ? typeName : bin}" accepts: ${Object.keys(ValueTypes.nameIndex)}. Leave empty for auto-detect.`);
```

Reply:
can't do the `const` bit because I also need new `binary` for later, the `[1]` part